File: Python/367. Valid Perfect Square.py
class Solution1:
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        """
        tmp = int(math.sqrt(num))
        return tmp*tmp == num

class Solution2:
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        """
        if num == 1:
            return True
        l, r = 0, num
        while l < r:
            mid = l + (r - l) // 2
            if mid * mid == num:
                return True
            elif mid * mid > num:
                r = mid
            else:
                l = mid + 1
        return False

File: C++/367. Valid Perfect Square.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        long long l = 0,r=num;
        if(num==1)
            return true;
        while(l<r)
        {
            long long mid = l + (r-l)/2;
            if(mid*mid == num)
                return true;
            else if(mid*mid < num)
                l = mid + 1;
            else
                r = mid;
        }
        return false;
    }
};